### PR TITLE
Reworked MaxText+TE integration

### PR DESCRIPTION
**Pull request by jberchtold-nvidia**
# Description

A new integration of TE+MaxText utilizing what we've learned in the initial integration to make this integration minimally invasive and easier to rebase while still maintaining the existing functionality. Most of this integration works directly as part of MaxText's own quantization system in `quantizations.py` so it will automatically be quantizing the correct tensors and leaving others, e.g. logits, unquantized in the same way as the base MaxText quantization operates.

This also avoids initialization of params in TE layers themselves and relies on MaxText's layers for param initialization and sharding and calls TE's function API instead of TE's layers. This avoids additional logic that would include more code changes and likely more conflicts when rebasing.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [ ] I have run end-to-end tests tests and provided workload links above if applicable.
- [ ] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/layers/normalizations.py [hunk 1]
```diff
@@ -56,6 +56,20 @@ def __init__(
 
   def __call__(self, x: jnp.ndarray) -> jnp.ndarray:
     """Applies layer normalization on the input."""
+    use_te_norm = True # TODO: get access to config to determine if TE norm should be enabled
+    if use_te_norm:
```
Comment (jberchtold-nvidia) [inline MaxText/layers/normalizations.py:61]: `rms_norm` in MaxText is called without the `config` so it's not as easy to toggle TE's norm here.

I'm thinking we have a few possible options here. These options only affect the pre-attention norm, the LNMLP norm is addressed already by using TE's `layernorm_mlp` call (assuming `fused_mlp` is enabled)

1. Provide the config as an additional argument. This would allow us to introduce new config options for toggling TE norm, but would require updating every callsite of `rms_norm` so would make this harder to rebase in future.
2. Do not integrate TE norm here. JAX norm is on-par or faster in our cases so far, so we can use MaxText's impl for now.
3. Integrate TE norm here and control it's enablement via an environment variable.

I'm leaning towards 2 or 3, more so 2 as I think that will be easier to rebase and upstream. But lmk what you think

Comment (phu0ngng) [reply MaxText/layers/normalizations.py:61]: I think we should do (3). 
The norm integration will mainly be for our internal testing (if any), as there are no performance benefits now. 
I don't see its urge to make the codebase more complicated to go for (1), but at the same time, it is good to have the integration done if we ever need to test it. 
I think we can have it off by default.

File: MaxText/layers/linears.py [hunk 2]
```diff
@@ -426,19 +443,21 @@ def __call__(self, inputs, decode: bool = False, deterministic: bool = False):
...
-        inputs = nn.with_logical_constraint(inputs, ("activation_batch",
-                                                     "prefill_activation_norm_length",
-                                                     "activation_embed")
-                                            )
-      else:
-        inputs = nn.with_logical_constraint(inputs, ("activation_batch",
-                                                     "activation_norm_length",
-                                                     "activation_embed")
-                                            )
+      inputs = nn.with_logical_constraint(inputs, dot_1_input_axes)
```
Comment (phu0ngng) [inline MaxText/layers/linears.py:460]: You may want to rebase your branch, as there was a fix for these sharding constraints after you cherry-pick the PR.

Comment (phu0ngng) [inline MaxText/layers/linears.py:460]: You may restore this change to minimize the line change ;)

Comment (jberchtold-nvidia) [reply MaxText/layers/linears.py:460]: Wouldn't that result in specifying the axes twice, once for TE via dot_1_input_axes above and then again here for MaxText? I'd coupled them so TE is always getting the same axes, but if we want a smaller diff I can undo this coupling

Comment (phu0ngng) [reply MaxText/layers/linears.py:460]: I see. I somehow missed that part :D. 
All good now.

File: MaxText/layers/normalizations.py (hunk 1 above)
Comment (jberchtold-nvidia) [reply MaxText/layers/normalizations.py:61]: I've updated to (3) with an env var in the latest version

**Review by phu0ngng — APPROVED** [review]
LGTM. Let's merge the PR!
